### PR TITLE
[SPARK-30386][SQL] Pushdown limit to outer join through projects

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Now LimitPushDown just push down limit if OUTER JOIN is child of LocalLimit. It used in such scence `df1.join(df2, LeftOuter).limit(1)`. 
We can also push down through Project if Project are deterministic like `df1.join(df2, LeftOuter).select("c").limit(1)`.

### Why are the changes needed?

Optimize sql.
```
create table test (c string);
explain extended select t1.c from test t1 left join test t2 on t1.c = t2.c limit 1;

--before
GlobalLimit 1
+- LocalLimit 1
   +- Project [c#5]
      +- Join LeftOuter, (c#5 = c#12)
         :- Relation[c#5]
         +- Filter isnotnull(c#12)
            +- Relation[c#12]

-- after
GlobalLimit 1
+- LocalLimit 1
   +- Project [c#35]
      +- Join LeftOuter, (c#35 = c#36)
         :- LocalLimit 1
         :  +- Relation[c#35]
         +- Filter isnotnull(c#36)
            +- Relation[c#36]
```

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Add UT.

### CODE REVIEW

**Comment by ulysses-you** [comment]
The failed test seems not caused by this pr.
cc @maropu @cloud-fan

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -495,6 +495,18 @@ object LimitPushDown extends Rule[LogicalPlan] {
         case _ => join
       }
       LocalLimit(exp, newJoin)
+    // Pushdown limit to outer join through projects if projects are deterministic.
+    case l @ LocalLimit(exp, proj @ Project(projs, join @ Join(left, right, joinType, _, _))) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: how about limit -> project -> union?

**Comment by HyukjinKwon** [comment]
@ulysses-you, seems the test failure is related and persistent:

```
Error Message
org.scalatest.exceptions.TestFailedException: transform should fail when ids exceed integer range. Expected exception org.apache.spark.SparkException to be thrown, but no exception was thrown
Stacktrace
sbt.ForkMain$ForkError: org.scalatest.exceptions.TestFailedException: transform should fail when ids exceed integer range. Expected exception org.apache.spark.SparkException to be thrown, but no exception was thrown
	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:530)
	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:529)
	at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1560)
	at org.scalatest.Assertions.intercept(Assertions.scala:824)
	at org.scalatest.Assertions.intercept$(Assertions.scala:804)
	at org.scalatest.FunSuite.intercept(FunSuite.scala:1560)
	at org.apache.spark.ml.recommendation.ALSSuite.testTransformIdExceedsIntRange$1(ALSSuite.scala:698)
	at org.apache.spark.ml.recommendation.ALSSuite.$anonfun$new$75(ALSSuite.scala:711)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.scalatest.Assertions.withClue(Assertions.scala:1223)
	at org.scalatest.Assertions.withClue$(Assertions.scala:1210)
	at org.scalatest.FunSuite.withClue(FunSuite.scala:1560)
	at org.apache.spark.ml.recommendation.ALSSuite.$anonfun$new$65(ALSSuite.scala:695)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:186)
	at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:149)
	at org.scalatest.FunSuiteLike.invokeWithFixture$1(FunSuiteLike.scala:184)
	at org.scalatest.FunSuiteLike.$anonfun$runTest$1(FunSuiteLike.scala:196)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:286)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: It pushed.  When execute `explain extended select c from test union all select c from test limit 1`, the optimized plan is 
```
GlobalLimit 1
+- LocalLimit 1
   +- Union
      :- LocalLimit 1
      :  +- Relation[c#50] parquet
      +- LocalLimit 1
         +- Relation[c#50] parquet
```

The difference is another optimizer `PushProjectionThroughUnion` has pushed down `Project`.

**Comment by ulysses-you** [comment]
@HyukjinKwon oh I see...
The reason is this UT use `DataFrame.first()` to do action and check if the total DataFrame satisfy the condition.

I think it can be instead of `DataFrame.collect()` ?

**Comment by ulysses-you** [comment]
Sorry I made a mistake. Union push limit in subquery but not through project. We can close this pr. Thanks for the review.

**Pull request closed without merging**